torch/distributed/distributed_c10d.py

```diff
@@ -1297,6 +1298,65 @@ def get_node_local_rank(fallback_rank: Optional[int] = None) -> int:
     )
 
 
+def _get_backends_for_timeout_set(group: ProcessGroup) -> Set[c10d_Backend]:
+    """
+    Return the all possible backends for a given process group when we
+    need to set the timeout for a ProcessGroup.
+
+    Args:
+        group (ProcessGroup): The process group to be set timeout.
+
+    Returns:
+        set[c10d_Backend]: A set of backends which we can set timeout for.
+    """
+    assert isinstance(group, ProcessGroup)
+    devices = group._device_types
+    backends = set()
+    if torch.device("cpu") in devices and is_gloo_available():
+        backend = group._get_backend(torch.device("cpu"))
+        if isinstance(backend, ProcessGroupGloo):
+            backends.add(backend)
+    if torch.device("cuda") in devices:
+        backend = group._get_backend(torch.device("cuda"))
+        if is_nccl_available() and isinstance(backend, ProcessGroupNCCL):
+            backends.add(backend)  # type: ignore[arg-type]
+        elif is_gloo_available() and isinstance(backend, ProcessGroupGloo):
+            backends.add(backend)  # type: ignore[arg-type]
+    if len(backends) == 0:
+        warnings.warn("Set timeout is now only supported for either nccl or gloo.")
+    return backends  # type: ignore[return-value]
+
+
+def _extend_timeout_all_pgs(timeout: timedelta) -> None:
+    """
+    This API extends the timeout for all PGs locally on one rank.
+    If user sets a very large negative timeout, we will set the timeout
+    to be the negative value.
+    NOTE: We only support to set timeout for cuda and gloo backends.
+
+    Args:
+        timeout (timedelta): The delta of timeout to extend.
+
+    Returns:
+        None.
+    """
+    zero_delta = timedelta()
+    for pg in _world.pg_map.keys():
+        backends = _get_backends_for_timeout_set(pg)
+        for backend in backends:
+            current_timeout = backend.options._timeout
+            if timeout + current_timeout > zero_delta:
+                backend._set_default_timeout(timeout + current_timeout)
+            else:
+                reset_timeout = (
+                    default_pg_nccl_timeout
+                    if torch.device("cuda") in pg._device_types
+                    and default_pg_nccl_timeout
+                    else default_pg_timeout
+                )
+                backend._set_default_timeout(reset_timeout)
+
+
 def _set_pg_timeout(timeout: timedelta, group: Optional[ProcessGroup] = None) -> None:
```

Comment:
I'm not really sure why we need to offer a new api. Users could directly call set_pg_timeout and modify it to a value they want. 

Can they also get the current timeout? Then they can do all the crazy math and bookkeeping.

Reply:
I guess the extra we provide here is let users to set timeout for all subPGs and do we want users to iterate through PGs themselves?